meteor/client/ui/Settings/ShowStyle/VariantSettings.tsx

```diff
@@ -266,225 +121,59 @@ export const ShowStyleVariantsSettings = withTranslation()(
-				collect: (monitor: DropTargetMonitor) => ({
-					isOver: monitor.isOver(),
-					canDrop: monitor.canDrop(),
-				}),
-			})
-
+		private provideShowStyleVariantItem = (variant: ShowStyleVariant, index: number) => {
 			return (
-				<table ref={drop} className={className}>
-					{children}
-				</table>
+				<VariantListItem
+					key={unprotectString(variant._id)}
+					index={index}
+					showStyleVariant={variant}
+					showStyleBase={this.props.showStyleBase}
+					dndVariants={this.props.showStyleVariants}
+					blueprintConfigManifest={this.props.blueprintConfigManifest}
+					t={this.props.t}
+					i18n={this.props.i18n}
+					tReady={this.props.tReady}
+				></VariantListItem>
 			)
 		}
 
-		VariantContainer = () => {
-			const moveVariantHandler = useCallback((dragIndex: number, hoverIndex: number) => {
-				const prevState = this.state.dndVariants
-				this.setState({
-					dndVariants: update(prevState, {
-						$splice: [
-							[dragIndex, 1],
-							[hoverIndex, 0, prevState[dragIndex] as ShowStyleVariant],
-						],
-					}),
-				})
-			}, [])
-
-			const returnVariantsForList = () => {
-				return this.state.dndVariants.map((variant: ShowStyleVariant, index: number) => (
-					<this.VariantItem
-						key={unprotectString(variant._id)}
-						index={index}
-						showStyleVariant={variant}
-						moveVariantHandler={moveVariantHandler}
-					></this.VariantItem>
-				))
-			}
-
+		render() {
 			const { t } = this.props
 			return (
 				<div>
 					<h2 className="mhn">{t('Show Style Variants')}</h2>
-					<DndProvider backend={HTML5Backend}>
-						<div>
-							<this.VariantList className="table expando settings-studio-showStyleVariants-table">
-								{returnVariantsForList()}
-							</this.VariantList>
-						</div>
-						<div className="mod mhs">
-							<button className="btn btn-primary" onClick={this.onAddShowStyleVariant}>
-								<FontAwesomeIcon icon={faPlus} />
-							</button>
-							<button className="btn btn-secondary mls" onClick={this.downloadAllShowStyleVariants}>
-								<FontAwesomeIcon icon={faDownload} />
-								&nbsp;{t('Export')}
-							</button>
-							<UploadButton
-								className="btn btn-secondary mls"
-								accept="application/json,.json"
-								onChange={(event) => this.importShowStyleVariants(event)}
-								key={this.state.timestampedFileKey}
-							>
-								<FontAwesomeIcon icon={faUpload} />
-								&nbsp;{t('Import')}
-							</UploadButton>
-							<button className="btn btn-secondary right" onClick={this.confirmRemoveAllShowStyleVariants}>
-								<FontAwesomeIcon icon={faTrash} />
-							</button>
-						</div>
-					</DndProvider>
+					<div>
+						<DndListWrapper
+							className="table expando settings-studio-showStyleVariants-table"
+							onDrop={this.persistStateVariants}
+							dndType={DragDropItemTypes.VARIANT}
+							stateUIArray={this.state.dndVariants}
+							provideItemForRender={this.provideShowStyleVariantItem}
```

Comment:
Just asking out of curiosity: Why did you choose this way of providing the items instead of just passing them as children?

Reply:
I tried that at first, but had issues with the dataflow. I must admit that i can't exactly remember what the issues were right now... :)

Reply:
I had this thought too. As far as I could see (I might be wrong) you need to rerender the item components with new keys everytime you drag them. If you parse them as children then you would not update the keys because you are basically moving an "already rendered" component because the component is rendered "outside" of the component handling the ordering.

Reply:
Yours might actually be the right solution. It looks similar to this https://beta.reactjs.org/reference/react/cloneElement#passing-data-with-a-render-prop, which is shown as an alternative to what I thought previously was the right way (iterating and cloning the children)